pkg/operator/etcdcertsigner/etcdcertsignercontroller.go

```diff
@@ -202,41 +241,35 @@ func (c *EtcdCertSignerController) sync(ctx context.Context, syncCtx factory.Syn
 	return updateErr
 }
 
-func (c *EtcdCertSignerController) syncAllMasterCertificates(ctx context.Context, recorder events.Recorder) error {
-	// TODO(thomas): it is of utmost importance to keep the existing signer certs for now
-	// when we just create a new signer cert, the new revision does not allow the peer to join the existing two-node
-	// cluster based on the old CA. Any newly rotated additions will come for free through the signerCaBundle and will work out of the box.
-	// Unfortunately, we can't make use of the certrotation.RotatedSigningCASecret here because it would immediately try to rotate the existing signer.
-	signerCaPair, err := tlshelpers.ReadConfigSignerCert(ctx, c.secretClient)
-	if err != nil {
-		return err
-	}
-
-	c.reportExpirationMetric(signerCaPair, "signer-ca")
-
-	// TODO(thomas): we need to transition that new signer as a replacement for the above - today we only bundle it
-	newSignerCaPair, _, err := c.certConfig.signerCert.EnsureSigningCertKeyPair(ctx)
+func (c *EtcdCertSignerController) syncAllMasterCertificates(
+	ctx context.Context, recorder events.Recorder, forceSkipRollout bool, lastRotationRevision int32, currentRevision int32) error {
+	signerCaPair, _, err := c.certConfig.signerCert.EnsureSigningCertKeyPair(ctx)
 	if err != nil {
 		return fmt.Errorf("error on ensuring etcd-signer cert: %w", err)
 	}
+	c.reportExpirationMetric(signerCaPair, "signer-ca")
 
-	metricsSignerCaPair, err := tlshelpers.ReadConfigMetricsSignerCert(ctx, c.secretClient)
+	metricsSignerCaPair, _, err := c.certConfig.metricsSignerCert.EnsureSigningCertKeyPair(ctx)
 	if err != nil {
-		return err
+		return fmt.Errorf("error on ensuring metrics-signer cert: %w", err)
 	}
-
 	c.reportExpirationMetric(metricsSignerCaPair, "metrics-signer-ca")
 
-	// TODO(thomas): we need to transition that new signer as a replacement for the above - today we only bundle it
-	newMetricsSignerCaPair, _, err := c.certConfig.metricsSignerCert.EnsureSigningCertKeyPair(ctx)
+	signerBundle, metricsSignerBundle, rolloutTriggered, err := c.ensureBundles(ctx, recorder, signerCaPair, metricsSignerCaPair, currentRevision)
 	if err != nil {
-		return fmt.Errorf("error on ensuring metrics-signer cert: %w", err)
+		return fmt.Errorf("error on ensuring bundles: %w", err)
 	}
 
-	signerBundle, metricsSignerBundle, err := c.ensureBundles(ctx, recorder,
-		[]*crypto.CA{signerCaPair, newSignerCaPair}, []*crypto.CA{metricsSignerCaPair, newMetricsSignerCaPair})
-	if err != nil {
-		return fmt.Errorf("error on ensuring bundles: %w", err)
+	if !forceSkipRollout {
+		if rolloutTriggered {
+			klog.Infof("skipping EtcdCertSignerController leaf cert generation as revision rollout has been triggered")
+			return nil
+		}
+
+		if currentRevision <= lastRotationRevision {
+			klog.Infof("skipping EtcdCertSignerController leaf cert generation as safe revision is not yet achieved, currently at %d - rotation happend at %d", currentRevision, lastRotationRevision)
```

Comment:
Isn't `currentRevision` the latest available revision that all nodes have settled on (i.e no rollout in progress).
In that case, if `currentRevision = lastRotationRevision` then isn't it safe to proceed.

But I guess setting a new `lastRotationRevision` will always trigger a new `currentRevision` so we always expect `currentRevision>lastRotationRevision` if the rollout has stabilized.

Reply:
This is a useful log line but I'm wondering if its ever even printed. Because we already skip during rollouts at L213 and see:
`skipping EtcdCertSignerController can't get current revision. Err=revision rollout in progress, can't establish current revision`

and once that's done `currentRevision>lastRotationRevision` so we're good to go again.

I should probably check the CI runs and stop speculating :)

Reply:
it is printed, but for a very short period of time, basically until the watch cache updates the CEO/etcd CRD